File: framework/scala/src/main/scala/soda/scala/unittest/conv/NestedIntegerConverter.scala
package soda.scala.unittest.conv

import play.api.libs.json._
import soda.scala.leetcode.NestedInteger

class NestedIntegerConverter extends ObjectConverter[NestedInteger] {

  override def fromJsonSerializable(js: JsValue): NestedInteger = NestedIntegerConverter.parse(js)

  override def toJsonSerializable(element: NestedInteger): JsValue = NestedIntegerConverter.serialize(element)

}

object NestedIntegerConverter {
  def parse(obj: JsValue): NestedInteger = {
    obj match {
      case i: JsNumber => {
        val ni = new NestedInteger {}
        ni.setInteger(i.as[Int])
        ni
      }
      case arr: JsArray => {
        val ni = new NestedInteger {}
//        val ni = DefaultNestedInteger.forList()
        for (elem <- arr.value) {
          ni.add(parse(elem))
        }
        ni
      }
      case _ => throw new RuntimeException("[NestedInteger] error type of object: " + obj.getClass)
    }
  }

  def serialize(ni: NestedInteger): JsValue = {
    if (ni.isInteger) Json.toJson(ni.getInteger) else Json.toJson(ni.getList.map(serialize))
  }
}

File: framework/kotlin/src/main/kotlin/soda/kotlin/unittest/conv/NestedIntegerConverter.kt
package soda.kotlin.unittest.conv

import kotlinx.serialization.json.*
import soda.kotlin.leetcode.NestedInteger

object NestedIntegerConverter {
    fun parse(obj: JsonElement): NestedInteger {
        return when (obj) {
            is JsonPrimitive -> NestedInteger(obj.int)
            is JsonArray -> {
                val ni = NestedInteger()
                for (elem in obj.jsonArray) {
                    ni.add(parse(elem))
                }
                ni
            }
            else -> throw RuntimeException("[NestedInteger] error type of object: ${obj::class}")
        }
    }

    fun serialize(ni: NestedInteger): JsonElement {
        return if (ni.isInteger()) {
            Json.encodeToJsonElement(ni.getInteger())
        } else {
            Json.encodeToJsonElement(ni.getList()?.map(this::serialize))
        }
    }

    fun create(): ObjectConverter<NestedInteger> {
        return ObjectConverter(this::parse, this::serialize)
    }
}